File: src/dotnet/FaxCore.ApiSamples/Samples/User/Message/SendMessage.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using System.Web.Script.Serialization;

namespace FaxCore.ApiSamples.Samples.User.Message
{
    internal static class SendMessage
    {
        public static async Task<string> RunAsync(FaxCoreClient client, SampleConfig config)
        {
            if (string.IsNullOrWhiteSpace(config.UploadFile))
            {
                throw new InvalidOperationException("Set FAXCORE_UPLOAD_FILE or FaxCoreUploadFile before running user.message.send. The send sample uploads this file first, then sends it as a message document.");
            }

            if (!File.Exists(config.UploadFile))
            {
                throw new FileNotFoundException("Upload file was not found.", config.UploadFile);
            }

            var uploadResponse = await client.UploadFileAsync("/api/upload", config.UploadField, config.UploadFile, config.UploadContentType).ConfigureAwait(false);
            var documents = new List<object>();

            if (!string.IsNullOrWhiteSpace(config.CoverPageName))
            {
                documents.Add(new
                {
                    name = config.CoverPageName,
                    path = string.Empty,
                    isMerge = true
                });
            }

            documents.Add(ExtractUploadedDocument(uploadResponse));
            var agents = BuildAgents(config);

            var request = new
            {
                message = new
                {
                    recipients = new[]
                    {
                        new
                        {
                            name = "Sample Recipient",
                            address = "+15551234567",
                            rawFax = true,
                            notifyAddress = "+15551234567",
                            company = "Example Company"
                        }
                    },
                    senderName = "FaxCore API Sample",
                    senderCompName = "Example Company",
                    subject = "FaxCore API sample fax",
                    note = "Sent from the .NET Framework sample.",
                    billingCode = "Sample",
                    scheduleDate = DateTime.Today.ToString("yyyy-MM-dd"),
                    priority = 60,
                    isOnHold = false,
                    mss = false,
                    msf = false,
                    trackings = new[]
                    {
                        new { label = "Sample", value = ".NET Framework" }
                    },
                    documents = documents.ToArray(),
                    agents = agents
                }
            };

            var sendResponse = await client.PostAsync("/api/message/send", request).ConfigureAwait(false);
            return "Upload response: " + uploadResponse + Environment.NewLine + "Send response: " + sendResponse;
        }

        private static object[] BuildAgents(SampleConfig config)
        {
            if (string.IsNullOrWhiteSpace(config.AgentID) ||
                string.IsNullOrWhiteSpace(config.AgentType) ||
                string.IsNullOrWhiteSpace(config.AgentValue))
            {
                return new object[0];
            }

            return new object[]
            {
                new
                {
                    id = config.AgentID,
                    type = config.AgentType,
                    value = config.AgentValue
                }
            };
        }

        private static object ExtractUploadedDocument(string uploadResponse)
        {
            var serializer = new JavaScriptSerializer();
            var value = serializer.DeserializeObject(uploadResponse);
            var root = value as IDictionary;
            var data = root != null && root.Contains("data") ? root["data"] as IEnumerable : null;

            if (data != null)
            {
                foreach (var item in data)
                {
                    var file = item as IDictionary;
                    if (file == null)
                    {
                        continue;
                    }

                    var id = file.Contains("id") ? file["id"] as string : null;
                    var fileName = file.Contains("fileName") ? file["fileName"] as string : null;
                    if (!string.IsNullOrWhiteSpace(id) && !string.IsNullOrWhiteSpace(fileName))
                    {
                        return new
                        {
                            name = id,
                            path = fileName,
                            isMerge = false
                        };
                    }
                }
            }

            throw new InvalidOperationException("Upload response did not include data[0].id and data[0].fileName: " + uploadResponse);
        }
    }
}

File: src/javascript/user/message/send-message.mjs
import { readFile, stat } from 'node:fs/promises';
import { basename } from 'node:path';
import { FaxCoreClient, printResult } from '../../common/faxcoreClient.mjs';

const uploadFile = process.env.FAXCORE_UPLOAD_FILE;
if (!uploadFile) {
  throw new Error('FAXCORE_UPLOAD_FILE is required. The send sample uploads this file first, then sends it as a message document.');
}

await stat(uploadFile);

const client = new FaxCoreClient();

const formData = new FormData();
const contentType = process.env.FAXCORE_UPLOAD_CONTENT_TYPE || guessContentType(uploadFile);
const file = contentType ? new Blob([await readFile(uploadFile)], { type: contentType }) : new Blob([await readFile(uploadFile)]);
formData.append(process.env.FAXCORE_UPLOAD_FIELD || '', file, basename(uploadFile));

const uploadResult = await client.postForm('/api/upload', formData);
const uploadedDocument = extractUploadedDocument(uploadResult);
const documents = [uploadedDocument];
const agents = buildAgents();

if (process.env.FAXCORE_COVER_PAGE_NAME) {
  documents.unshift({
    name: process.env.FAXCORE_COVER_PAGE_NAME,
    path: '',
    isMerge: true
  });
}

const sendResult = await client.post('/api/message/send', {
  message: {
    recipients: [
      {
        name: process.env.FAXCORE_RECIPIENT_NAME || 'Sample Recipient',
        address: process.env.FAXCORE_RECIPIENT_FAX || '+15551234567',
        rawFax: true,
        notifyAddress: process.env.FAXCORE_NOTIFY_ADDRESS || process.env.FAXCORE_RECIPIENT_FAX || '+15551234567',
        company: process.env.FAXCORE_RECIPIENT_COMPANY || 'Example Company'
      }
    ],
    senderName: process.env.FAXCORE_SENDER_NAME || 'FaxCore API Sample',
    senderCompName: process.env.FAXCORE_SENDER_COMPANY || 'Example Company',
    subject: process.env.FAXCORE_SUBJECT || 'FaxCore API sample fax',
    note: process.env.FAXCORE_NOTE || 'Sent from the JavaScript sample.',
    billingCode: process.env.FAXCORE_BILLING_CODE || 'Sample',
    scheduleDate: process.env.FAXCORE_SCHEDULE_DATE || today(),
    priority: Number(process.env.FAXCORE_PRIORITY || 60),
    isOnHold: false,
    mss: false,
    msf: false,
    trackings: [
      { label: 'Sample', value: 'JavaScript' }
    ],
    documents,
    agents
  }
});

printResult('Upload file', uploadResult);
printResult('Send message', sendResult);

function extractUploadedDocument(value) {
  const status = typeof value?.status === 'string' ? value.status.toLowerCase() : 'success';
  if (status !== 'success') {
    throw new Error(`Upload failed: ${value?.message || JSON.stringify(value)}`);
  }

  const item = Array.isArray(value?.data) ? value.data[0] : undefined;
  if (item?.id && item?.fileName) {
    return {
      name: item.id,
      path: item.fileName,
      isMerge: false
    };
  }

  throw new Error(`Upload response did not include data[0].id and data[0].fileName: ${JSON.stringify(value)}`);
}

function guessContentType(path) {
  const lower = path.toLowerCase();
  if (lower.endsWith('.pdf')) return 'application/pdf';
  if (lower.endsWith('.tif') || lower.endsWith('.tiff')) return 'image/tiff';
  if (lower.endsWith('.png')) return 'image/png';
  if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg';
  if (lower.endsWith('.txt')) return 'text/plain';
  if (lower.endsWith('.doc')) return 'application/msword';
  if (lower.endsWith('.docx')) return 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
  return '';
}

function today() {
  return new Date().toISOString().slice(0, 10);
}

function buildAgents() {
  const id = process.env.FAXCORE_AGENT_ID;
  const type = process.env.FAXCORE_AGENT_TYPE;
  const value = process.env.FAXCORE_AGENT_VALUE;
  return id && type && value ? [{ id, type, value }] : [];
}
